File: gajeong/Programmers/LV2/부대복귀.js
function solution(n, roads, sources, destination) {
  var answer = [];
  const map = Array.from(Array(n + 1), () => []);

  for (let i = 0; i < roads.length; i++) {
    let [a, b] = roads[i];
    map[a].push(b);
    map[b].push(a);
  }

  // 목적지에서 각 점으로 도달할 수 있는지 반대로 계산하면 됨.
  let start = destination;

  const visited = Array(n + 1).fill(false);
  const dist = Array(n + 1).fill(-1);
  visited[start] = true;
  dist[start] = 0;
  const queue = [[start, 0]];

  let idx = 0;
  while (queue.length > idx) {
    //종료조건

    let [now, d] = queue[idx];

    for (let k = 0; k < map[now].length; k++) {
      let nxt = map[now][k];
      if (visited[nxt]) continue;
      queue.push([nxt, d + 1]);
      visited[nxt] = true;
      dist[nxt] = d + 1;
    }

    ++idx;
  }

  sources.forEach((s) => {
    answer.push(dist[s]);
  });

  return answer;
}
console.log(
  solution(
    3,
    [
      [1, 2],
      [2, 3],
    ],
    [2, 3],
    1
  )
);

File: jiwoo/Programmers/부대복귀.cpp
#include <bits/stdc++.h>

using namespace std;

int vis[100001];
int de[100001];
vector<int> ro[100001];

vector<int> solution(int n, vector<vector<int>> roads, vector<int> sources, int destination)
{
    vector<int> answer;

    for (int i = 0; i < roads.size(); i++)
    {
        ro[roads[i][0]].push_back(roads[i][1]);
        ro[roads[i][1]].push_back(roads[i][0]);
    }

    queue<pair<int, int>> q;
    
    q.push({destination, 0});
    vis[destination] = 1;

    while (!q.empty())
    {
        int cur = q.front().first;
        int dest = q.front().second;
        q.pop();
        de[cur] = dest;

        for (int i = 0; i < ro[cur].size(); i++)
        {
            int next = ro[cur][i];
            if (vis[next] == 0)
            {
                q.push({next, dest + 1});
                vis[next] = 1;
            }
        }
    }

    for (int i = 0; i < sources.size(); i++)
    {
        if (sources[i] == destination)
        {
            answer.push_back(0);
            continue;
        }
        if (de[sources[i]] == 0)
            answer.push_back(-1);
        else
            answer.push_back(de[sources[i]]);
    }

    return answer;
}

int main()
{
    ios::sync_with_stdio(0);
    cin.tie(0);

    return 0;
}